### DISCUSSION TITLE
Adding a Divisional Coupler

### DISCUSSION THREAD

**Discussion by Psi58** [GrandOrgue/grandorgue#1690] [Help] [answered]
I am trying, and failing, to add a Divisional Coupler to an old format sampleset (Friesach). My console has pistons under the keyboards, but not over the pedals, so my intention is for this coupler to allow the manual pistons to activate both the manual and  pedal divisional combinations.

The requirement is not unlike [Discussion #121](https://github.com/GrandOrgue/grandorgue/discussions/121), which I have followed carefully.

I have successfully (I think...) added the coupler to the left jamb, and ODFEdit finds no errors. But when I select my new coupler and press a combination in the Divisionals panel, GO always returns an "Unhandled Exception" error.

I wonder if anyone is aware of an organ that contains a Divisional Coupler that I can use as a template for my ODF (none of the organs I have looked at do)?

Any help or advice appreciated.

**Comment by Psi58** [comment]
I still haven't cracked this, so here's a bit more information in case anyone can see what I'm doing wrong.
I've simplified things by abandoning the attempt to put this coupler on the left jamb (for now), and reverted to placing it on the main panel. I've created new 'Coupler7.png' graphics.

Here are the key parts of my ODF:

[Organ]
...
NumberOfDivisionalCouplers=1
NumberOfSwitches=55
...

[DivisionalCoupler001]
Name=Gt & Pedal Pistons Coupler
NumberOfManuals=2
Manual001=001
Manual002=000
DefaultToEngaged=N
Displayed=N
BiDirectionalCoupling=Y
SwitchCount=1
Switch001=055

[Switch055]
Name=Gt & Pedal Pistons Coupler
DefaultToEngaged=N
GCState=0
StoreInDivisional=N
StoreInGeneral=N
Displayed=Y
DispLabelText=
PositionX=0
PositionY=67
ImageOn=Data - Friesach\Images\Console\0\On\Coupler7.png
ImageOff=Data - Friesach\Images\Console\0\Off\Coupler7.png
MouseRadius=0
MouseRectLeft=7
MouseRectWidth=134
MouseRectTop=7
MouseRectHeight=132

This produces no errors in ODFedit, and 2 warnings in GO, 
Warning: Unused ODF entry 'DivisionalCoupler001/Switch001'
Warning: Unused ODF entry 'DivisionalCoupler001/SwitchCount'
If engaged, the coupler doesn't couple the Divisionals from the respective manuals.

As the above warnings suggested that the separate switch might be an issue, I experimented by combining the presentation with the functional definition:

[Organ]
...
NumberOfDivisionalCouplers=1
NumberOfSwitches=54
...

[DivisionalCoupler001]
Name=Gt & Pedal Pistons Coupler
NumberOfManuals=2
Manual001=001
Manual002=000
DefaultToEngaged=N
Displayed=Y
BiDirectionalCoupling=Y
GCState=0
StoreInDivisional=N
StoreInGeneral=N
DispLabelText=
PositionX=0
PositionY=67
ImageOn=Data - Friesach\Images\Console\0\On\Coupler7.png
ImageOff=Data - Friesach\Images\Console\0\Off\Coupler7.png
MouseRadius=0
MouseRectLeft=7
MouseRectWidth=134
MouseRectTop=7
MouseRectHeight=132

Again, there are no errors in ODFedit, and this time no warnings from GO.
With the coupler off, the Divisionals work fine for their respective manuals. But once the coupler is on then any piston causes the 'Unhandled Exception' error as described earlier.

Suggestions and comments very welcome...

**Reply by Psi58** [reply]
I'm beginning to think this might simply be a bug.

Adding the Divisional Coupler in GOODF resulted in the same 'Unhandled Exception' error. 

I'll wait a few days to see if anyone has any insight and, if not, will raise it as an issue.

In the meantime, I'm still keen to examine and learn from any samplesets that have Divisional Couplers, so please share details of any candidates.

**Reply by larspalo** [reply, marked as answer]
@Psi58 @oleg68 Yes, this is a bug in GrandOrgue. When the Divisional Coupler has BiDirectionalCoupling=Y set then GrandOrgue will crash when you first activate the divisional coupler and then push a divisional. If BiDirectionalCoupling=N is set everything works as it should. Feel free to test the attached sample set that demonstrates it. The Ped/Gt Div Cplr works as it's not bi directional vs the Gt/Sw will crash GO.

[TestDivisionalCoupler.zip](https://github.com/GrandOrgue/grandorgue/files/13404753/TestDivisionalCoupler.zip)

Change line 103 to BiDirectionalCoupling=N and see that it works (upwards).

Expected behavior is that GO should not crash at all and that if the BiDirectionalCoupling=Y coupler is used the coupling of the divisionals should work in both directions for Gt/Sw in this case.

**Reply by Psi58** [reply]
Many thanks. Now raised as an Issue, so will mark this discussion as answered.
